### ISSUE TITLE
XR status is not updated if XR is patched from composition functions

### ISSUE THREAD

**Issue by mproffitt** [crossplane/crossplane#4968]
When running compositions in pipeline mode, an issue arises as a result of the patching carried out on the XR as part of the compoosition function runner here: 

https://github.com/crossplane/crossplane/blob/603e4ca7c85b851e5088469942b48ab020431cf8/internal/controller/apiextensions/composite/composition_functions.go#L391-L393

When the patch is applied, the XR `resourceVersion` field is updated on the managed resource but is never reflected back into the XR as it's handed back to the main reconile loop.

The outcome of this is that when crossplane tries to update the conditions inside the reconciler here

https://github.com/crossplane/crossplane/blob/603e4ca7c85b851e5088469942b48ab020431cf8/internal/controller/apiextensions/composite/reconciler.go#L711

it is unable to proceed with the error being masked inside the `reconcile.Result`.

The error returned from the client here is `the object has been modified; please apply your changes to the latest version and try again` which seems to get lost by the client logs making this difficult to trace.

This manifests inside the XR as "Stuck" status fields similar to that reported in slack here: https://crossplane.slack.com/archives/CEG3T90A1/p1699029022000269 despite their underlying resources being marked ready, for example when I run my pipeline, most of the resources become ready except 2 of them:

- eks-cluster-vpc (vpcs.ec2.aws.upbound.io)
- awsmanagedcontrolplane (objects.kubernetes.crossplane.io)

In the crossplane beta trace tree, these objects show as `ready` but unready against the XR:

```
$ crossplane beta trace xi sample-customer-zkrp6
NAME                                                 SYNCED   READY   STATUS
CompositeImport/sample-customer-zkrp6                True     False   Unready resources: awsmanagedcontrolplane, eks-cluster-vpc
...
├─ VPC/sample-customer                               True     True    Available
├─ ClusterAuth/sample-customer-eks-cluster-auth      True     True    Available
...
├─ Object/sample-customer-awsmanagedcontrolplane     True     True    Available
...
```

Another side effect I've seen from this is composition function error status being reflected from transient timeouts that don't seem to clear, for example if a running function temporarily became incommunicado, e.g. through pod restart, this presents as an rpc error even if the pod is subsequently running correctly.

```
  conditions:
  - lastTransitionTime: "2023-11-04T14:34:22Z"
    message: 'cannot compose resources: cannot run Composition pipeline step "generate-subnets":
      cannot run Function "function-generate-subnets": rpc error: code = DeadlineExceeded
      desc = context deadline exceeded'
    reason: ReconcileError
    status: "False"
    type: Synced
  - lastTransitionTime: "2023-11-03T07:16:09Z"
    message: 'Unready resources: awsmanagedcontrolplane, eks-cluster-vpc'
    reason: Creating
    status: "False"
    type: Ready
```

Examining both the function pod logs and the crossplane pod logs shows the function working and behaving normally.

To verify the behaviour I was seeing, I looked at modifying the behaviour of the reconciler at the end of the `Reconcile` function to retrieve the last applied `resourceVersion` and this seems to have a positive effect:

```golang
	nxr := composite.New(composite.WithGroupVersionKind(r.gvk))
	if err := r.client.Get(ctx, req.NamespacedName, nxr); err != nil {
		log.Debug(errGet, "error", err)
		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGet)
	}
	
	xr.SetResourceVersion(nxr.GetResourceVersion())
	if err = r.client.Status().Update(ctx, xr); err != nil {
		log.Debug(errUpdateStatus, "error", err)
	}
	return reconcile.Result{RequeueAfter: r.pollInterval}, errors.Wrap(err, errUpdateStatus)
```

Whilst re-fetching the XR is neither a clean or elegant solution, I wasn't able to determine what the new `resourceVersion` would be without doing so

```
$ crossplane beta trace xi sample-customer-zkrp6
NAME                                                 SYNCED   READY   STATUS
CompositeImport/sample-customer-zkrp6                True     True    Available
...
├─ VPC/sample-customer                               True     True    Available
...
├─ Object/sample-customer-awsmanagedcontrolplane     True     True    Available
...
```

**Comment by negz** [comment]
We have [an E2E test](https://github.com/crossplane/crossplane/blob/v1.14.0/test/e2e/apiextensions_test.go#L187) that verifies that it's possible to patch XR status using functions. The XR status is back-propagated to the claim. So it seems like this is working in at least some situations.

> When the patch is applied, the XR resourceVersion field is updated on the managed resource but is never reflected back into the XR as it's handed back to the main reconile loop.

When you say managed resource here, do you mean the XR?

I would have expected `client.Status().Patch()` to load the updated resource version into `XR`. 🤔 That seems to be happening in our E2E tests:

```
XR "apiextensions-composition-functions-kj2df" revision BEFORE COMPOSE: "980"
2023-11-04T23:34:27Z    DEBUG   crossplane      Reconciling     {"controller": "offered/compositeresourcedefinition.apiextensions.crossplane.io", "request": {"name":"xnopresources.nop.example.org"}, "uid": "61860f6d-b38f-4dc0-b8f7-7047e5a4583b", "version": "708", "name": "xnopresources.nop.example.org", "controller": "claim/xnopresources.nop.example.org", "request": {"name":"apiextensions-composition-functions","namespace":"default"}}
XR "apiextensions-composition-functions-kj2df" revision AFTER status PATCH: "1018"
2023-11-04T23:34:27Z    DEBUG   crossplane      Reconciling     {"controller": "defined/compositeresourcedefinition.apiextensions.crossplane.io", "controller": "composite/xnopresources.nop.example.org", "request": {"name":"apiextensions-composition-functions-kj2df-v4rxj"}}
2023-11-04T23:34:27Z    DEBUG   crossplane      Pipeline step "be-a-dummy": I am doing a compose!       {"controller": "defined/compositeresourcedefinition.apiextensions.crossplane.io", "controller": "composite/xnopresources.nop.example.org", "request": {"name":"apiextensions-composition-functions-kj2df"}, "uid": "3b21b0ce-6a8a-4ab8-9b7e-d6ab950742c8", "version": "980", "name": "apiextensions-composition-functions-kj2df"}
2023-11-04T23:34:27Z    DEBUG   crossplane      Composed resource is not yet ready      {"controller": "defined/compositeresourcedefinition.apiextensions.crossplane.io", "controller": "composite/xnopresources.nop.example.org", "request": {"name":"apiextensions-composition-functions-kj2df"}, "uid": "3b21b0ce-6a8a-4ab8-9b7e-d6ab950742c8", "version": "980", "name": "apiextensions-composition-functions-kj2df", "id": "nested-xr"}
XR "apiextensions-composition-functions-kj2df" revision AFTER COMPOSE: "1018"
```

<details>

```diff
diff --git a/internal/controller/apiextensions/composite/composition_functions.go b/internal/controller/apiextensions/composite/composition_functions.go
index d006fec4..b1dc87ce 100644
--- a/internal/controller/apiextensions/composite/composition_functions.go
+++ b/internal/controller/apiextensions/composite/composition_functions.go
@@ -391,6 +391,7 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
        if err := c.client.Status().Patch(ctx, xr, client.Apply, client.ForceOwnership, client.FieldOwner(FieldOwnerXR)); err != nil {
                return CompositionResult{}, errors.Wrap(err, errApplyXRStatus)
        }
+       fmt.Printf("XR %q revision AFTER status PATCH: %q\n", xr.GetName(), xr.GetResourceVersion())

        // Produce our array of resources to return to the Reconciler. The
        // Reconciler uses this array to determine whether the XR is ready.
diff --git a/internal/controller/apiextensions/composite/reconciler.go b/internal/controller/apiextensions/composite/reconciler.go
index a7dcc0b8..7a164aed 100644
--- a/internal/controller/apiextensions/composite/reconciler.go
+++ b/internal/controller/apiextensions/composite/reconciler.go
@@ -614,6 +614,8 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
                return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
        }

+       fmt.Printf("XR %q revision BEFORE COMPOSE: %q\n", xr.GetName(), xr.GetResourceVersion())
+
        // TODO(negz): Pass this method a copy of xr, to make very clear that
        // anything it does won't be reflected in the state of xr?
        res, err := r.resource.Compose(ctx, xr, CompositionRequest{Revision: rev, Environment: env})
@@ -628,6 +630,8 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
                return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
        }

+       fmt.Printf("XR %q revision AFTER COMPOSE: %q\n", xr.GetName(), xr.GetResourceVersion())
+
        if r.kindObserver != nil {
                var gvks []schema.GroupVersionKind
                for _, ref := range xr.GetResourceReferences() {
@@ -704,6 +708,8 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
                return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
        }

+       fmt.Printf("XR %q revision BEFORE status UPDATE: %q\n", xr.GetName(), xr.GetResourceVersion())
+
        // We requeue after our poll interval because we can't watch composed
        // resources - we can't know what type of resources we might compose
        // when this controller is started.
```

</details>

> which seems to get lost by the client logs making this difficult to trace

As of v1.14 we intentionally don't surface these issues as Kubernetes events because they're almost always transient and should clear up on the next reconcile.

> To verify the behaviour I was seeing, I looked at modifying the behaviour of the reconciler at the end of the Reconcile function to retrieve the last applied resourceVersion and this seems to have a positive effect

Unfortunately we can't do this, as it could lead to data loss. The revision is used for optimistic concurrency. We want updates to fail if the client as an outdated revision, otherwise the update could potentially reset the status to an outdated revision.

Can you share your XRD, XR, and Composition to help reproduce this?

**Comment by negz** [comment]
Just thinking out loud: If something else (the claim controller?) was updating the XR so frequently that its resource version was constantly changing between when the function composer patched the status and when the reconciler tried to update the status that could cause the behavior seen here.

**Comment by negz** [comment]
I just tried to reproduce this using the the following manifests. No luck so far.

https://github.com/negz/function-patch-and-transform/tree/repro-cc-4968/example

**Comment by mproffitt** [comment]
> Can you share your XRD, XR, and Composition to help reproduce this?

Of course. Everything is contained on the `composition-rework` branch at https://github.com/giantswarm/crossplane-cluster-import/tree/composition-rework/crossplane - I'm only focusing on AWS at the moment though so you're free to ignore anything for other providers.

- The relevant Composition: https://github.com/giantswarm/crossplane-cluster-import/blob/composition-rework/crossplane/composition/composition-aws.yaml
- The associated XRD: https://github.com/giantswarm/crossplane-cluster-import/blob/composition-rework/crossplane/composition/composition-aws.yaml
- The sanitised XR: https://github.com/giantswarm/crossplane-cluster-import/blob/composition-rework/xrender/observed-xr.yaml

This composition uses two custom functions. They are:

- crossplane-fn-generate-subnets: https://github.com/giantswarm/crossplane-fn-generate-subnets/
- crossplane-fn-describe-nodegroups: https://github.com/giantswarm/crossplane-fn-describe-nodegroups

> When you say managed resource here, do you mean the XR?

In this instance yes. I was trying to differentiate between the XR loaded into memory inside the controller and the XR persisted inside the cluster. Potentially a poor choice of words though.

> As of v1.14 we intentionally don't surface these issues as Kubernetes events because they're almost always transient and should clear up on the next reconcile.

I absolutely agree about this not being raised as an event. As a cluster admin, this would be noise. Are these logged though? I never saw them inside the crossplane logs and it was only when I specifically logged the error that I began to understand why I was seeing this behaviour. (I would be willing to admit I probably didn't look very hard here though).

To try and explain a little further. When I first saw this issue, I did wonder if it was a result of a dependency chain.

If we look at the resources, I first create a Cluster object with `managementPolicies: ["Observe"]`.  The VPC object is also created but cannot complete because the Cluster has not reconciled its status back to the XR here: https://github.com/giantswarm/crossplane-cluster-import/blob/composition-rework/crossplane/composition/composition-aws.yaml#L123 which is read by the VPC here: https://github.com/giantswarm/crossplane-cluster-import/blob/composition-rework/crossplane/composition/composition-aws.yaml#L138

The cluster object is also used by the `generate-subnets` step which is a custom function that exists here: https://github.com/giantswarm/crossplane-fn-generate-subnets/

This function is the only one of the two additional functions that actually writes back to the XR (here: https://github.com/giantswarm/crossplane-fn-generate-subnets/blob/main/fn.go#L115) although I do not persist this change but instead hand that back for crossplane to persist for me.

What I find interesting about this is only a few of the 15 resources created by this composition actually fail. Most are updated on the status correctly. However in the XR above, only two of the listed "waiting" resources persistently fail for me.

**Comment by mproffitt** [comment]
> If something else (the claim controller?) 

Given what I've come to understand about the crossplane machinery, I'm not seeing evidence of this and to the contrary, on the status of the assosiated claim, I see these conditions:

```yaml
 conditions:
  - lastTransitionTime: "2023-11-05T11:59:02Z"
    reason: ReconcileSuccess
    status: "True"
    type: Synced
  - lastTransitionTime: "2023-11-05T11:59:02Z"
    message: Composite resource claim is waiting for composite resource to become
      Ready
    reason: Waiting
    status: "False"
    type: Ready
```

This would imply to me that the claim controller isn't actually doing anything right now as it's waiting for the XR to complete?

**Comment by mproffitt** [comment]
I tried a varient of your patch above with additional print in the `claim controller` and what I'm seeing is that the claim is applied twice for each reconciliation loop

```
crossplane-6d65f78db8-rtwb5 crossplane ===== LOADING COMPOSITE RESOURCE =====
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE COMPOSE: "34904"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "34904"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "34904"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE status PATCH: ""
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision AFTER status PATCH: "34969"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "34969"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "34984"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision AFTER COMPOSE: "34969"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE status UPDATE: "34969" -- AT CLUSTER: "34984"
crossplane-6d65f78db8-rtwb5 crossplane ===== LOADING COMPOSITE RESOURCE =====
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE COMPOSE: "34984"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE status PATCH: ""
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision AFTER status PATCH: "35098"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "35098"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "35102"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision AFTER COMPOSE: "35098"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE status UPDATE: "35098" -- AT CLUSTER: "35102"
crossplane-6d65f78db8-rtwb5 crossplane ===== LOADING COMPOSITE RESOURCE =====
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE COMPOSE: "35102"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE status PATCH: ""
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision AFTER status PATCH: "35208"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "35208"
crossplane-6d65f78db8-rtwb5 crossplane CLAIM ON XR "sample-customer-45ptp" revision: BEFORE APPLY: "35217"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision AFTER COMPOSE: "35208"
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE status UPDATE: "35208" -- AT CLUSTER: "35217"
crossplane-6d65f78db8-rtwb5 crossplane ===== LOADING COMPOSITE RESOURCE =====
crossplane-6d65f78db8-rtwb5 crossplane XR "sample-customer-45ptp" revision BEFORE COMPOSE: "35217"
```

The first apply uses the same XR `resourceVersion` as the `composite controller` whilst the second gives the version which blocks the update.

<details>

```diff
diff --git a/internal/controller/apiextensions/claim/reconciler.go b/internal/controller/apiextensions/claim/reconciler.go
index f4fb4430..0942e2a2 100644
--- a/internal/controller/apiextensions/claim/reconciler.go
+++ b/internal/controller/apiextensions/claim/reconciler.go
@@ -19,6 +19,7 @@ package claim

 import (
        "context"
+       "fmt"
        "time"

        "github.com/google/go-cmp/cmp"
@@ -505,6 +506,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
                return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, cm), errUpdateClaimStatus)
        }

+       fmt.Printf("CLAIM ON XR %q revision: BEFORE APPLY: %q\n", cp.GetName(), cp.GetResourceVersion())
        err := r.client.Apply(ctx, cp, resource.AllowUpdateIf(func(old, obj runtime.Object) bool { return !cmp.Equal(old, obj) }))
        switch {
        case resource.IsNotAllowed(err):
diff --git a/internal/controller/apiextensions/composite/composition_functions.go b/internal/controller/apiextensions/composite/composition_functions.go
index d006fec4..cba381b9 100644
--- a/internal/controller/apiextensions/composite/composition_functions.go
+++ b/internal/controller/apiextensions/composite/composition_functions.go
@@ -388,9 +388,11 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
        xr.SetKind(k)
        xr.SetName(n)

+       fmt.Printf("XR %q revision BEFORE status PATCH: %q\n", xr.GetName(), xr.GetResourceVersion())
        if err := c.client.Status().Patch(ctx, xr, client.Apply, client.ForceOwnership, client.FieldOwner(FieldOwnerXR)); err != nil {
                return CompositionResult{}, errors.Wrap(err, errApplyXRStatus)
        }
+       fmt.Printf("XR %q revision AFTER status PATCH: %q\n", xr.GetName(), xr.GetResourceVersion())

        // Produce our array of resources to return to the Reconciler. The
        // Reconciler uses this array to determine whether the XR is ready.
diff --git a/internal/controller/apiextensions/composite/reconciler.go b/internal/controller/apiextensions/composite/reconciler.go
index a7dcc0b8..abbeb872 100644
--- a/internal/controller/apiextensions/composite/reconciler.go
+++ b/internal/controller/apiextensions/composite/reconciler.go
@@ -480,6 +480,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
        ctx, cancel := context.WithTimeout(ctx, timeout)
        defer cancel()

+       fmt.Printf("===== LOADING COMPOSITE RESOURCE =====\n")
        xr := composite.New(composite.WithGroupVersionKind(r.gvk))
        if err := r.client.Get(ctx, req.NamespacedName, xr); err != nil {
                log.Debug(errGet, "error", err)
@@ -614,6 +615,8 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
                return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
        }

+       fmt.Printf("XR %q revision BEFORE COMPOSE: %q\n", xr.GetName(), xr.GetResourceVersion())
+
        // TODO(negz): Pass this method a copy of xr, to make very clear that
        // anything it does won't be reflected in the state of xr?
        res, err := r.resource.Compose(ctx, xr, CompositionRequest{Revision: rev, Environment: env})
@@ -628,6 +631,8 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
                return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
        }

+       fmt.Printf("XR %q revision AFTER COMPOSE: %q\n", xr.GetName(), xr.GetResourceVersion())
+
        if r.kindObserver != nil {
                var gvks []schema.GroupVersionKind
                for _, ref := range xr.GetResourceReferences() {
@@ -708,6 +713,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
        // resources - we can't know what type of resources we might compose
        // when this controller is started.
        xr.SetConditions(xpv1.Available())
+
+       nxr := composite.New(composite.WithGroupVersionKind(r.gvk))
+       if err := r.client.Get(ctx, req.NamespacedName, nxr); err != nil {
+               log.Debug(errGet, "error", err)
+               return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGet)
+       }
+
+       fmt.Printf("XR %q revision BEFORE status UPDATE: %q -- AT CLUSTER: %q\n", xr.GetName(), xr.GetResourceVersion(), nxr.GetResourceVersion())
        return reconcile.Result{RequeueAfter: r.pollInterval}, errors.Wrap(r.client.Status().Update(ctx, xr), errUpdateStatus)
 }
```
</details>

**Comment by negz** [comment]
@mproffitt Could you try create just the XR directly, without a claim, and see whether Crossplane can update its status?

**Comment by mproffitt** [comment]
That seems to have an effect:

```nohighlight
$ k get importclaims.crossplane.giantswarm.io -A
No resources found
$ k get ximports.crossplane.giantswarm.io
NAME                    SYNCED   READY   COMPOSITION      AGE
sample-customer-45ptp   True     True    aws-eks-import   2m42s
$ k get xi sample-customer-45ptp -o yaml | yq .status.conditions
- lastTransitionTime: "2023-11-05T19:28:25Z"
  reason: ReconcileSuccess
  status: "True"
  type: Synced
- lastTransitionTime: "2023-11-05T19:29:20Z"
  reason: Available
  status: "True"
  type: Ready
```

**Comment by negz** [comment]
Thank you, that's useful. It certainly seems to be a race between the claim and composite controllers.

What's interesting is that in our E2E tests the XR and claim eventually do become ready. I wonder what the difference is. Is there any chance that your XR spec or status is never reaching a steady state? Is the function pipeline causing it to change each time the XR is reconciled

> Are these logged though? I never saw them inside the crossplane logs and it was only when I specifically logged the error that I began to understand why I was seeing this behaviour. (I would be willing to admit I probably didn't look very hard here though).

It doesn't seem like they are. We only recently refactored to suppress these - I agree they should be debug logged.

**Comment by mproffitt** [comment]
> Is there any chance that your XR spec or status is never reaching a steady state

Not that I've observed but this did put me in mind of some additional tests from my own functions.

In the first test, I removed the patching from `function-generate-subnets` and updated the composition to reflect that `status.subnets` wouldn't be available. In this instance, everything became ready.

In the second test, I updated the function to only add the subnets if the number of subnets matches those found on the VPC - in this instance, I witnessed the same behaviour of the XR never becoming ready.

In the third test, I added a random patch to the output of the `function-describe-subnets` to see if that would have an effect. It didn't and the XR still remained in an unready condition.

In the final test, I updated the `function-generate-subnets` function to specifically sort the `status.subnets` field by subnet ID before adding it to the XR. Equivalently, this has also had no effect on the XR which remains unready. However, this would have no effect on the order after being marshalled and sent over `gRPC` so the order can't (or shouldn't) be the problem.

Out of these 4 tests, only not adding the patch has succeeded but interestingly, when I tested earlier without the claim in place, the XR did succeed despite the same functions running in the same way.

Once the claim is applied, this is never updated, at least in my local `kind` environment. On my remote cluster, the claim is managed by flux but that wouldn't have an effect unless the `metadata` or `spec` of the claim changed, not the status.

**Comment by negz** [comment]
Hrm, so far I'm stumped. I opened https://github.com/crossplane/crossplane/pull/4971 to try to reproduce this but it seems to work fine.

Could you try add the `x-kubernetes-list-type` and `x-kubernetes-list-map-keys` OpenAPI extensions to `status.subnets` per https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy (see #4971 for an example). Crossplane uses server-side apply to apply a function pipeline's desired state, and this should allow it to intelligently merge entries in the subnets array rather than replacing the whole array atomically each time. I don't really expect this to fix the issue, but it's a good practice either way so worth a shot.

**Comment by mproffitt** [comment]
I'm going to say that's a really good call. It did fix the issue but more, it made me re-examine that section of the XRD which didn't have an entry for the `tags` object. This is what I think must have been really causing the issue.

The original was:

```yaml
  subnets:
    type: array
    items:
      type: object
      additionalProperties: true
```
which has been replaced with (as a bare minimal set)

```yaml
  subnets:
  type: array
  x-kubernetes-list-type: map
  x-kubernetes-list-map-keys:
  - id
  items:
    type: object
    x-kubernetes-map-type: granular
    additionalProperties: true
    properties:
      id:
        description: The ID for this subnet
        type: string
      tags:
        type: object
        x-kubernetes-map-type: granular
        additionalProperties: true
    required:
    - id
```

If I may, this is something that could be well served in a troubleshooting section of your documentation, that if resources don't become ready, to check the definition of what's being returned.

So what's the real issue here? The issue is that I'm returning more information from my custom function than crossplane understands how to apply, mostly as a result of the over generalisation of the original XRD spec (`additionalProperties: true` when the function result may contain sub objects or arrays)

Whilst I don't particularly see this as being a crossplane problem specifically, I could easily imagine this being a pain point for others, especially in larger organisations where function creation and XRD specification might be carried out by different teams, or even when consuming functions from the community.

Here, the problem manifested as a bug in crossplane when in fact it really wasn't and was a result of my own poorly defined XRD spec.

Out of curiosity, I looked more at the errors that are masked from the `patch` in the composition functions, the `apply` in the claim and the `update` in the composition reconcilers and as far as I can see, there is no indication from any of them of a mismatched object tree.

In fact the only indication I can find of something going wrong is that final suppressed error from the update 

```nohighlight
the object has been modified; please apply your changes to the latest version and try again
```

One final thought from me around function consumption in a pipeline; `input` to a function is a defined statement "If your function accepts input, this is how you specify it". Output from a function is generally undefined as many functions such as `function-go-template` or `function-cue` are more freeform.

`function-patch-and-transform` again fits into the generic model but only operates on the patches defined in composition however as highlighted here, this leaves a gap in how crossplane is capable of handling and as demonstrated here, if a function patches the XR, it's capable of breaking the state if its patch definition doesn't match the XRD definition.

This leads me to wonder if there is a way of instructing crossplane that "My function is going to patch the XR, this is how it's going to do it" and if that patch doesn't match the XRD at that location, the patch fails?
